### PR TITLE
Switch to Keras Autodoc

### PR DESCRIPTION
**Pull request by leonoverweel**
This PR replaces our fork of `pydoc-markdown` with `keras-autodoc`, which should simplify our setup a bit. It also moves our top-of-module markdown-in-python docs to the format that keras-autodoc prefers: markdown templates where the API docs for the module are placed at the `{{autogenerated}}` tag.

There are some regressions that need to be fixed before merging. Probably non-exhaustive list:
- [x] `larq.models.summary()` doesn't get parsed -> this is because of the `print` function we pass; by refactoring that slightly I can fix it.
- [x] Docstrings across Larq, Zoo, and Compute Engine need to be updated to follow the [Google Python style guide](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#381-docstrings) to be rendered correctly. Currently some things are slightly different (mostly spacing). We also need to delete the introduction markdown strings, since those are in templates now.
  - [x] larq/larq#482
    - [x] (and merge)
  - [x] larq/zoo#170
    - [x] (and merge)
  - [x] larq/compute-engine#343
    - [x] (and merge)
- [x] Introductions are missing (e.g. `larq.layers` only has docstrings for the classes, but not the intro text)
- [x] Order of classes is different
- [x] The fancy view source button is uglified (don't think this is blocking)

### CODE REVIEW

**Comment by AdamHillier** [comment]
Arguments and input/output shape seem a bit broken: image

**Comment by leonoverweel** [comment]
@AdamHillier yep that's because of this:

> - [ ]  Docstrings across Larq, Zoo, and Compute Engine need to be updated to follow the Google Python style guide to be rendered correctly. Currently some things are slightly different (mostly spacing). We also need to delete the introduction markdown strings, since those are in templates now.

**Comment by lgeiger** [comment]
Just for reference, the typehints are not rendered correctly: https://docs-4o0q1xndv.now.sh/larq/api/optimizers/#caseoptimizer-class

**Comment by leonoverweel** [comment]
**Waiting with this for now.** The type hints getting rendered as part of the signatures is a blocking regression (e.g. [old](https://docs.larq.dev/larq/api/optimizers/#caseoptimizer) vs. [new](https://docs-4o0q1xndv.now.sh/larq/api/optimizers/#caseoptimizer-class) since it makes the docs less readable. Will revisit depending on keras-team/keras-autodoc#71.

**Comment by leonoverweel** [comment]
Testing keras-team/keras-autodoc#78 which adds replacing long ugly TF types with shorter types, on top of keras-team/keras-autodoc#77 which moved type hints out of signatures and into docstrings. 

image

@lgeiger thoughts on this?

**Comment by lgeiger** [comment]
I don't have strong opinions on moving the type hints out of signatures or keeping them in.

However I don't think it is nice to have the signature in one line so one needs to scroll to see the defaults. In my opinion this becomes very unreadable, especially when having a lot of arguments like the layers (e.g. see the signature of `Dense` in https://keras.io/layers/core/)

**Comment by leonoverweel** [comment]
Yep the signatures are actually Black-formatted, but at a max line length of 110 which happens to make this one one line. I'm hoping we [can make this configurable](https://github.com/keras-team/keras-autodoc/pull/78#issuecomment-617846927).

**Comment by lgeiger** [comment]
> Yep the signatures are actually Black-formatted, but at a max line length of 110 which happens to make this one one line.

Ah OK, if the docstrings of the layers look fine I'm not too bothered by a bit of scrolling for small functions

**Comment by leonoverweel** [comment]
Looks like this is failing and I'll have to branch off a different commit of LCE (whichever was the last pip release) and add these changes on top of that, but that's a small issue and beyond that this is ready for re-review.

**Comment by lgeiger** [comment]
Thanks for the changes, I took a look at the generated site and found a few regressions compared to master:
- [ ] References in [`QuantConv2DTranspose`](https://docs-git-keras-autodoc.plumerai-larq.now.sh/larq/api/layers/#quantconv2dtranspose-class) break site navigation and table of contents:
Screenshot 2020-05-07 at 15 56 20
- [ ] Code links loose styling and are now displayed as `[source]`. I personaly don't mind this for now and we can fix this in a future PR
- [ ] Code links are not aligned with the heading:
Screenshot 2020-05-07 at 16 00 36
- [ ] Code links link to github master instead of linking to the relevant tag. This can cause link breakage when files on GitHub change.
- [ ] Unnecessary `class` and `function` statements in headings which add clutter in the table of contents:
Screenshot 2020-05-07 at 16 03 16
- [ ] Type annotation in [`modes.summary`](https://docs-git-keras-autodoc.plumerai-larq.now.sh/larq/api/models/#summary-function) mentions `tensorflow.python.keras.engine.training.Model` instead of `tf.keras.Model`:
Screenshot 2020-05-07 at 16 05 58

Other than that I really like the new look of the API docs.

File: requirements.txt [hunk 1]
```diff
@@ -2,14 +2,15 @@ tensorflow==2.1.0
...
 pymdown-extensions==7.1
 git+git://github.com/lgeiger/mknotebooks@text-output-code
 mkdocs-minify-plugin==0.3.0
-larq-zoo==1.0.1
+git+git://github.com/larq/zoo@docs-refactor  # TODO: change back when released
 altair==4.1.0
 pandas==1.0.3
-larq-compute-engine==0.2.1
+larq-compute-engine==0.2.1  # TODO: bump when released
 git+https://github.com/lgeiger/pydoc-markdown.git
```
Comment (lgeiger) [inline requirements.txt:15]:
```suggestion
```

**Comment by lgeiger** [comment]
I found another visual change. Keyword only arguments are now prefixed with `*`:
Screenshot 2020-05-07 at 23 14 47
Is this intentional?

**Comment by leonoverweel** [comment]
It looks like the `*` [is also there in the code](https://github.com/larq/zoo/blob/v1.0.1/larq_zoo/literature/resnet_e.py#L146) so I think it's expected behavior; did we previously strip these out manually?

**Comment by lgeiger** [comment]
> It looks like the * is also there in the code so I think it's expected behavior; did we previously strip these out manually?

Sounds good, we can keep it consistent with the source code then.

**Comment by leonoverweel** [comment]
Cool, so I think once we merge larq/larq# and larq/zoo#170, I think we just need to figure out what releases (or master installs if we don't want to do a release) to point to in the reqs and then we should be ready to go.

**Comment by leonoverweel** [comment]
Ready to merge once:

- [ ] Larq is released with larq/larq#482 and we switch the dependency on this PR to that version
- [ ] Zoo is released with larq/zoo#170 and we switch the dependency on this PR to that version
- [ ] Compute Engine is released with larq/compute-engine#343 and we switch the dependency on this PR to that version

**Comment by gabrieldemarmiesse** [comment]
I'm glad that keras-autodoc helped you! Feel free to drop messages in our issue tracker to give feedback if you have any :)

**Comment by leonoverweel** [comment]
@gabrieldemarmiesse thanks! One thing that'd be nice is a version bump so that we can pull kerasa-autodoc from PyPI instead of a commit on GitHub.

**Comment by gabrieldemarmiesse** [comment]
I'll put that in my todo list :)

**Comment by gabrieldemarmiesse** [comment]
The new version is up on pypi! The documentation isn't updated yet, but you can look at the docstrings if needed for the moment.

**Pull request merged**